// avail-js-sdk/sdk/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SDK = exports.WaitFor = exports.Bytes = exports.Keyring = exports.BN = exports.sdkTransactionData = exports.sdkTransactions = void 0;
const chain_1 = require("../chain");
const transactions_1 = require("./transactions");
exports.sdkTransactions = __importStar(require("./transactions"));
exports.sdkTransactionData = __importStar(require("./transaction_data"));
var util_1 = require("@polkadot/util");
Object.defineProperty(exports, "BN", { enumerable: true, get: function () { return util_1.BN; } });
var api_1 = require("@polkadot/api");
Object.defineProperty(exports, "Keyring", { enumerable: true, get: function () { return api_1.Keyring; } });
var types_codec_1 = require("@polkadot/types-codec");
Object.defineProperty(exports, "Bytes", { enumerable: true, get: function () { return types_codec_1.Bytes; } });
var transactions_2 = require("./transactions");
Object.defineProperty(exports, "WaitFor", { enumerable: true, get: function () { return transactions_2.WaitFor; } });
class SDK {
    static New(endpoint) {
        return __awaiter(this, void 0, void 0, function* () {
            const api = yield (0, chain_1.initialize)(endpoint);
            return new SDK(api);
        });
    }
    constructor(api) {
        this.api = api;
        this.tx = new transactions_1.Transactions(api);
    }
}
exports.SDK = SDK;
